neuralmonkey/decoders/beam_search_decoder.py

```diff
@@ -248,65 +270,84 @@ def body(*args) -> BeamSearchLoopState:
             hyp_probs = tf.expand_dims(bs_state.logprob_sum, 1) + logprobs
 
             # update hypothesis lengths
             hyp_lengths = bs_state.lengths + 1 - tf.to_int32(bs_state.finished)
 
-            # shape(scores) = beam x vocabulary
+            # shape(scores) = (batch * beam) x vocabulary
             scores = hyp_probs / tf.expand_dims(
                 self._length_penalty(hyp_lengths), 1)
 
-            # flatten so we can use top_k
-            scores_flat = tf.reshape(scores, [-1])
+            # reshape to batch_size x (beam_size * vocabulary) for topk
+            input_beam_size = bs_state.input_beam_size
+            scores_flat = tf.reshape(
+                scores, [-1, input_beam_size * len(self.vocabulary)])
 
             # shape(both) = beam
             topk_scores, topk_indices = tf.nn.top_k(
                 scores_flat, self._beam_size)
-
-            topk_scores.set_shape([self._beam_size])
-            topk_indices.set_shape([self._beam_size])
+            topk_indices.set_shape([None, self._beam_size])
+
+            # first, we need offset to flatten topk indices
+            batch_offset = tf.expand_dims(
+                tf.range(
+                    0,
+                    self.batch_size * input_beam_size * len(self.vocabulary),
+                    input_beam_size * len(self.vocabulary)),
+                axis=1)
+            topk_indices_flat = tf.reshape(
+                topk_indices + batch_offset, [-1])
 
             # flatten the hypothesis probabilities
             hyp_probs_flat = tf.reshape(hyp_probs, [-1])
 
             # select logprobs of the best hyps (disregard lenghts)
-            next_beam_logprob_sum = tf.gather(hyp_probs_flat, topk_indices)
-            # pylint: disable=no-member
-            next_beam_logprob_sum.set_shape([self._beam_size])
-            # pylint: enable=no-member
-
-            next_word_ids = tf.mod(topk_indices,
-                                   len(self.vocabulary))
-
-            next_beam_ids = tf.div(topk_indices,
-                                   len(self.vocabulary))
-
-            next_finished = tf.gather(bs_state.finished, next_beam_ids)
-            next_just_finished = tf.equal(next_word_ids, END_TOKEN_INDEX)
+            next_beam_logprob_sum = tf.gather(hyp_probs_flat,
+                                              topk_indices_flat)
+
+            # next, we compute offset to flatten other beam_id indices
+            batch_offset = tf.expand_dims(
+                tf.range(
+                    0,
+                    self.batch_size * input_beam_size,
+                    input_beam_size),
+                axis=1)
+
+            next_word_ids = tf.mod(topk_indices, len(self.vocabulary))
+            next_word_ids_flat = tf.reshape(next_word_ids, [-1])
+            next_beam_ids = tf.div(topk_indices, len(self.vocabulary))
+            next_beam_ids_flat = tf.reshape(
+                next_beam_ids + batch_offset, [-1])
```

Comment:
`mod` a `div` dát k sobě, pak prázdnou řádku a pak ty flaty. To flatování je kravina, ty divy a mody musej bejt vidět a bejt u sebe

Reply:
OK